services/device-registry-base/src/main/java/org/eclipse/hono/service/management/credentials/CommonCredential.java

```diff
@@ -164,4 +167,46 @@ public void checkValidity() {
     public CommonCredential stripPrivateInfo() {
         return this;
     }
+
+    /**
+     * Merges the secrets of the given credential with that of the current one based on the secret ids.
+     *
+     * @param credential The credential to be merged.
+     * @return a reference to this for fluent use.
+     * @throws IllegalArgumentException if the given credential is invalid and cannot be merged.
+     * @throws NullPointerException if the given credential is {@code null}.
+     */
+    @JsonIgnore
+    public CommonCredential merge(final CommonCredential credential) {
+
+        Objects.requireNonNull(credential);
+
+        if (!getType().equals(credential.getType())) {
+            throw new IllegalArgumentException("credential to be merged must be of the same type");
+        }
+
+        getSecrets()
+                .forEach(secret -> Optional.ofNullable(secret.getId())
+                        .ifPresent(secretId -> findSecretById(credential, secretId)
+                                .map(result -> {
+                                    if (secret instanceof PasswordSecret) {
+                                        ((PasswordSecret) secret).merge((PasswordSecret) result);
+                                    } else if (secret instanceof PskSecret) {
+                                        ((PskSecret) secret).merge((PskSecret) result);
+                                    }
```

Comment:
why don't we simply define a merge method on `CommonSecret`?

Reply:
@sophokles73 I have added a merge method to the `CommonSecret`.